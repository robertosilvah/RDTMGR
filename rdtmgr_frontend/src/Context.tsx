import React, { useContext, useState, useEffect, useReducer, FC, Reducer, useCallback } from "react";

import useWebSocket from "react-use-websocket";

import { Filters } from "./Helpers";
import { LineProcess, LineProcesses, LineProcessIndicators, ShiftDefinition, Location, LineProcessState } from "./Models";
import { WebSocketHook } from "react-use-websocket/dist/lib/types";
import { CachePolicies, useFetch } from "use-http";
import { TimeFilterChangeEvent } from "./components/filters/TimeFilter";
import { GanttDatum } from "./components/reports/Gantt";

export type AppContextModel = {
	api?: {
		main?: string,
		webSocket?: string
	},
	clientId?: number,
	setState?: React.Dispatch<React.SetStateAction<AppContextModel>>
};

export const AppContext = React.createContext<AppContextModel>({});

AppContext.displayName = "AppContext";

export const AppContextProvider: FC = ({ children }) => {
	const [ state, setState ] = useState<AppContextModel>({
		api: {
			main: process.env.REACT_APP_API,
			webSocket: process.env.REACT_APP_WS
		}
	});

	useEffect(() => {
		setState(prev => {
			return { ...prev, setState };
		})
	}, []);

	return (
		<AppContext.Provider value={state}>
			{children}
		</AppContext.Provider>
	)
}

//#region FilterContext

export type TimeRange = [Date, Date];

export type FilterContextDefinition = {
	date?: Date,
	dateRange?: TimeRange,
	timeFilter?: string,
	shiftFilter?: string | number,
	weekFilter?: string,
	search?: string,
	shiftFilters?: ShiftDefinition[],
	setFilters?: React.Dispatch<TimeFilterChangeEvent>
};

export const FilterContext = React.createContext<FilterContextDefinition>({});

FilterContext.displayName = "FilterContext";

const defaultFilters = () => {
	const date = new Date();
	const timeFilter = "Shift";

	return {
		timeFilter,
		date,
	};
}

type FilterReducerFunction = (prev: FilterContextDefinition) => TimeFilterChangeEvent;
type FilterReducerActionHandler = (prev: FilterContextDefinition, payload: TimeFilterChangeEvent) => FilterContextDefinition;
type FilterReducer = Reducer<FilterContextDefinition, TimeFilterChangeEvent | FilterReducerFunction>;

type asd = keyof typeof Filters;
// const FilterActionHandlers: { [k in asd]: FilterReducerActionHandler } = {
// 	Day: (prev, payload) => {
// 		const {  } = prev;
// 	}
// };

const filterReducer: FilterReducer = (prev, filters) => {
	if (typeof filters === "function")
		filters = filters(prev);

	const { timeFilter, shiftFilter, dateRange, date } = filters;
	const state: FilterContextDefinition = {};

	if (dateRange) {
		state.date = date;
		state.dateRange = dateRange;
	} else {
		state.date = prev.date;
		state.dateRange = prev.dateRange;
	}

	state.timeFilter = timeFilter ?? prev.timeFilter;

	if (state.timeFilter === "Shift")
		state.shiftFilter = Number(shiftFilter ?? prev.shiftFilter);

	return state;
}

export const FilterProvider = ({children}) => {
	const [filters, setFilters] = useReducer(filterReducer, defaultFilters());
	const [state, setState] = useState({ setFilters });

	useEffect(() => setState({ ...filters, setFilters }), [filters]);

	return (
		<FilterContext.Provider value={state}>
			{children}
		</FilterContext.Provider>
	);
}

//#endregion

// #region Line process contexts

export const LineProcessesContext = React.createContext<LineProcesses>(null);

LineProcessesContext.displayName = "LineProcessesContext";

enum LineProcessEvents {
	OnFinishProduction = "OnFinishProduction",
	OnUpdateProduction = "OnUpdateProduction",
	OnFinishDelay = "OnFinishDelay",
	OnUpdateDelay = "OnUpdateDelay",
	OnStateUpdate = "OnStateUpdate",
	OnInitialization = "OnInitialization"
}

type LineProcessActionHandler = (prev: LineProcesses, payload: unknown) => LineProcesses;

type DelayUpdatedModel = {
	location: Location,
	indicators: LineProcessIndicators,
	currentDelay: GanttDatum;
};

type ProductionUpdatedModel = {
	location: Location,
	indicators: LineProcessIndicators,
	currentProduction: GanttDatum;
};

const LineProcessActionHandlers: { [k in LineProcessEvents]?: LineProcessActionHandler } = {
	OnInitialization: (_, payload: LineProcessState[]) => {
		const entries = payload.map<[number, LineProcessState]>(process => [process.location.id, process]);
		const state = new Map<number, LineProcessState>(entries);

		console.log("OnInitialization", state);

		return state;
	},

	OnStateUpdate: (prev, payload: LineProcess) => {
		const state = new Map(prev);
		const { location: { id }, shifts } = payload;

		console.log("OnStateUpdate", prev, payload);

		const process: LineProcessState = !state.has(id) ?
			{ ...payload, realtime: true, shifts } :
			{ ...state.get(id), ...payload };

		state.set(id, process);

		return state;
	},

	OnUpdateDelay: (prev, payload: DelayUpdatedModel) => {
		const states = new Map(prev);
		const { location: { id }, indicators, currentDelay } = payload;
		const prevState = states.get(id);

		if (!currentDelay)
			console.log("!currentDelay", payload);

		if (prevState) {
			const { gantt } = indicators;
			const state = { ...prevState, indicators: { ...indicators, gantt } };
			states.set(id, state);
		}

		return states;
	},

	OnUpdateProduction: (prev, payload: ProductionUpdatedModel) => {
		const states = new Map(prev);
		const { location: { id }, indicators } = payload;
		const prevState = states.get(id);

		if (prevState) {
			const { gantt } = indicators;
			const state: LineProcessState = { ...prevState, indicators: { ...indicators, gantt } };
			states.set(id, state);
		}

		return states;
	}
}

type LineProcessAction = {
	clientId?: number,
	type: LineProcessEvents,
	payload: unknown
};

const LineProcessesReducer: Reducer<LineProcesses, LineProcessAction | null | undefined> = (prev, action) => {
	if (!action)
		return prev;

	const { type, payload } = action;
	// console.log("LineProcessesReducer action:", action);

	if (type && payload)
		return LineProcessActionHandlers[type](prev, payload);

	return prev;
}

/** Retrieves the line processes current state and receives their realtime updates. */
export const LineProcessesProvider: FC = ({children}) => {
	const { api, setState } = useContext(AppContext);

	const [lineProcesses, dispatch] = useReducer(LineProcessesReducer, new Map());

	const { get } = useFetch<LineProcess[]>(`${api.main}/process`, null);

	const getProcesses = useCallback(async () => {
		const data = await get();

		if (data) {
			dispatch({
				type: LineProcessEvents.OnInitialization,
				payload: data
			})
		}
	}, []);

	useEffect(() => { getProcesses(); }, []);

	// establish websocket connection for realtime line process updates
	const {lastJsonMessage: lastMsg}: WebSocketHook<MessageEvent> = useWebSocket(api.webSocket, {
		onError: (e) => console.error(e),
		onOpen: (e) => {
			console.info("WebSocket connection opened");
		},
		onClose: (e) => {
			console.info("WebSocket connection closed");
		},
		shouldReconnect: () => true,
		retryOnError: true,
		reconnectAttempts: Infinity,
		reconnectInterval: 1000,
	});

	useEffect(() => {
		// console.log("lastMsg:", lastMsg);
		if (lastMsg) {
			const { clientId }: LineProcessAction = lastMsg;

			if (clientId && setState) {
				setState((prev) => {
					return { ...prev, clientId };
				})
			}

			dispatch(lastMsg);
		}
	}, [lastMsg]);

	return (
		<LineProcessesContext.Provider value={lineProcesses}>
			{children}
		</LineProcessesContext.Provider>
	);
}

export const LineProcessContext = React.createContext<LineProcessState>({ realtime: true });
LineProcessContext.displayName = "LineProcessContext";

type LineProcessProviderProps = {
	lineProcess?: Partial<LineProcess>;
}

export const LineProcessProvider: FC<LineProcessProviderProps> = ({lineProcess, children}) => {
	const { api, clientId } = useContext(AppContext);
	const { dateRange, shiftFilter, timeFilter, date } = useContext(FilterContext);

	const [realtime, setRealtime] = useState(true);
	const [state, setState] = useState<LineProcessState>({ realtime });

	const { post } = useFetch(`${api.main}/process/${lineProcess.location.id}`, { cachePolicy: CachePolicies.NO_CACHE });

	// updates the realtime state
	useEffect(() => {
		if (dateRange) {
			const now = new Date();
			const [start, end] = dateRange;

			setRealtime(now > start && now < end && timeFilter === "Shift");
			getProcess();
		}
	}, [timeFilter, shiftFilter, dateRange]);

	const getProcess = useCallback(async () => {
		const now = new Date();

		const [start, end] = dateRange;
		const [prevStart, prevEnd] = state.interval.range.map(s => new Date(s));
		const sameStart = start.getTime() === prevStart.getTime();
		const sameEnd = end.getTime() === prevEnd.getTime();

		if (sameStart && sameEnd)
			return;

		const body: {
			clientId: number,
			start?: Date | string,
			end?: Date | string,
			shift?: number | string,
			date?: Date
		} = { clientId };

		if (timeFilter === "Shift") {
			body.shift = shiftFilter;
			body.date = date;
		} else {
			body.start = start;
			body.end = end;
		}

		const result: LineProcess = await post(body);

		if (result) {
			const [start, end] = result.interval.range.map(s => new Date(s));
			const realtime = now > start && now < end && timeFilter === "Shift";
			setRealtime(realtime);
			setState(prev => {
				return { ...prev, ...result, realtime };
			});

			console.log("getProcess", shiftFilter, start, end, realtime);
		}
	}, [post, dateRange, realtime, shiftFilter, date, timeFilter, state]);

	useEffect(() => {
		setState((prev) => {
			if (realtime)
				return { ...lineProcess, realtime };

			return prev;
		});
	}, [lineProcess, realtime]);

	return (
		<LineProcessContext.Provider value={state}>
			{children}
		</LineProcessContext.Provider>
	);
}

// #endregion
