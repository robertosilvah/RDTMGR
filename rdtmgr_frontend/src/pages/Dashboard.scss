#prod-dashboard.report-content {
	margin: 16px;
	gap: 16px;
	grid-template-columns: repeat(7, 1fr);

	h3 {
		margin: 0;
		grid-column: 1 / span 7;
	}

	.filter-container {
		grid-column: 1 / span 7;
	}

	@media screen and (max-width: 1440px) {
		& {
			grid-template-columns: repeat(4, 1fr);
		}

		h3 {
			grid-column: 1 / span 4;
		}
	
		.filter-container {
			grid-column: 1 / span 4;
		}
	}
}

.asd {
	display: flex;
	grid-column: 1 / span 7;
	align-items: center;
	
	.mdc-select {
		margin-left: 8px;
		margin-right: 4px;
	}

	h3 span {
		color: var(--mdc-theme-text-secondary-on-background);
	}

	@media screen and (max-width: 1440px) {
		& {
			grid-column: 1 / span 4;
		}
	}
}