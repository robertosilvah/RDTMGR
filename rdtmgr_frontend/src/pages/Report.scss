.card-component.gantt {
	text-align: center;
}

tr > .k-header> .k-link {
	background-color: #787F83;
	color:white;
}

tr > .k-header> .k-link:hover {
	color:white;
}

tr > .k-header {
	background-color: #787F83;
}

.report-content {
	display: grid;
	margin: 16px;
	gap: 8px;
	grid-template-columns: 1fr;
	grid-template-rows: repeat(3, min-content) 1fr;
	// width: 100%;
	// width: calc(100% - 32px);

	.card-container {
		display: grid;
		margin: 0;
		gap: 8px;
		grid-column: 1 / -1;

		&.summary {
			grid-template-columns: 1fr 1fr 1fr;
			padding: 0;
			gap: 6px;

			.card-component {
				display: grid;
				grid-template-columns: repeat(8, auto);
				grid-template-rows: 1fr;
				gap: 16px;
				padding: 0 16px 8px;
				color: #5A5A5A;
			}
		}

		&.table {
			grid-column: 1 / -1;
			grid-row-end: -1;
			padding: 0;
			background: white;
		}

		.k-grid {
			border-width: 0;

			th, td {
				padding: 8px 16px;
			}
		}

		.card-toolbar {
			padding: 6px 16px;
			display: flex;
			grid-column: 1 / -1;
			border-bottom: solid green 2px;
			align-items: center;
		}
	}

	.summary-container {
		display: grid;
		margin: 0;
		grid-column: 1 / -1;

		.k-grid {
			border-radius: 0 0 4px 4px;
		}
	}

	.indicator {
		display: flex;
		flex-direction: column;

		.k-arcgauge-label {
			position: absolute;
			width: 100%;
			text-align: center;
		}

		&.center {
			align-items: center;
		}

		.title {
			position: relative;
			margin-bottom: 0px;
		}

		h4 {
			margin: 0;
		}
	}

	@media screen and (max-width: 1200px) {
		.card-container.summary .card-component {
			grid-template-columns: repeat(4, auto);
			grid-template-rows: auto auto;
		}
	}
}

.filter-container {
	display: flex;
	gap: 8px;
}