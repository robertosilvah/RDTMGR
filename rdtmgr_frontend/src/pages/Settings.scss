.rmwc-data-table__cell .mdc-text-field {
	vertical-align: middle;
	margin: -0.75rem -1rem;
	background-color: transparent;
	border-radius: 0;
	min-width: 160px;
}

.datetime-cell {
	display: flex;

	& > :nth-child(1) {
		margin-right: 0;
	}

	& > :nth-child(2) {
		margin-left: 0;
	}
}

.settings-page {
	display: grid;
	grid-template-rows: min-content 1fr;
	height: 100%;
	padding: 0 24px;

	@media screen and (min-width: 960px) {
		& {
			margin: 0 48px;
		}
	}

	@media screen and (min-width: 1280px) {
		& {
			margin: 0 96px;
		}
	}

	@media screen and (min-width: 1366px) {
		& {
			margin: 0 15%;
		}
	}

	.end {
		width: 100%;
		display: flex;
		justify-content: flex-end;
	}

	.table {
		margin-bottom: 24px;
	}
}
