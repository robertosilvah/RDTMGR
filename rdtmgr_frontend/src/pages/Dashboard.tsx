import React, { FC, Fragment, useCallback, useContext, useEffect, useState } from "react";

import moment from "moment";

import { CachePolicies, useFetch } from "use-http";

import { Select, Dialog, DialogTitle, DialogContent, DialogActions, DialogButton, DialogProps, TextField, IconButton } from "rmwc";

import { AppContext, FilterContext, FilterProvider, LineProcessContext, LineProcessesContext, LineProcessProvider, TimeRange } from "../Context";
import { IndicatorCard } from "../components/reports/IndicatorCard";
import { FullscreenSpinner } from "../components/Spinners";
import TimeFilter, { TimeInterval } from "../components/filters/TimeFilter";
import { LineProcess } from "../Models";
import { getShiftsStart, useFilterChangeCallback } from "../CustomHooks";

import "./Report.scss";
import "./Dashboard.scss";

// const handleMoreClick = () => console.log("More info clicked");

const calcColor = (value, standard?) => {
	switch (typeof value) {
		case "number":
			return value >= standard ? "#6dc78c" : "#ad4d4f";
		default:
			return value === standard ? "#6dc78c" : "#ad4d4f";
	}
}

const StandardSubtitle = ({text, stdValue}) => <>
	<span>{text}</span>
	<br/>
	<span>Standard: {stdValue}</span>
</>;

interface Product {
	id: number;
	name: string;
}

export const Dashboard = () => {
	const {api} = useContext(AppContext);
	const {date, timeFilter, setFilters} = useContext(FilterContext);
	const lineProcesses = useContext(LineProcessesContext);
	const {get} = useFetch(api.main, null);

	const [products, setProducts] = useState<Product[]>([]);

	const getProducts = useCallback(async () => {
		const result = await get("/products");
		if (result) setProducts(result);
	}, [get]);

	// get products
	useEffect(() => { getProducts() }, [getProducts]);

	return (
		<>
			<FullscreenSpinner/>

			<div id="prod-dashboard" className="report-content">
				{
					lineProcesses ?
					Array.from(lineProcesses.values()).map(process => (
						<FilterProvider key={process.location.name}>
							<LineProcessProvider lineProcess={process}>
								<LineState
									products={products}
								/>
							</LineProcessProvider>
						</FilterProvider>
					)) : "Loading..."
				}
			</div>
		</>
	);
}

type LineStateProps = {
	defaultState?: Partial<LineProcess>;
	products: Product[];
}

const LineState: FC<LineStateProps> = ({defaultState, products}) => {
	const { api } = useContext(AppContext);
	const { date, timeFilter, setFilters } = useContext(FilterContext);
	const lineProcess = useContext(LineProcessContext) || { realtime: true };

	const [state, setState] = useState<Partial<LineProcess>>(lineProcess);
	// const [realtime, setRealtime] = useState(true);
	const { location, realtime, shifts } = lineProcess;
	const [shift, setShift] = useState<TimeInterval>();

	const [open, setOpen] = useState(false);
	const [product, setProduct] = useState<Product>();
	const [productLock, setProductLock] = useState(true);

	const standard = lineProcess?.standard;

	const { post } = useFetch(`${api.main}/process`, { cachePolicy: CachePolicies.NO_CACHE });

	// updates the state and the date range if it changed
	useEffect(() => {
		if (lineProcess) {
			setState(prev => { return {...prev, ...lineProcess}; });
			
			if (realtime && timeFilter === "Shift") {
				if (shift?.id !== lineProcess.interval?.id) {
					const { interval } = lineProcess;
	
					if (interval) {
						console.log(`${lineProcess.location?.name}: updating shift to`, interval);
						const dateRange = interval.range.map(s => new Date(s)) as TimeRange;
	
						setFilters({ dateRange, shiftFilter: interval.id, date: getShiftsStart(shifts) });
						setShift(interval);
					}
				}
			}
		}
	}, [lineProcess, realtime, shifts]);

	useEffect(() => {
		if (standard && products) {
			if (!product || product.id !== standard.productId)
				setProduct(products.find(p => p.id === standard.productId));
		}
	}, [standard, products]);

	const handleAddClick = useCallback(async (amount) => {
		const locationId = location?.id;

		if (Number.isInteger(locationId))
			await post(`/${locationId}/scrap/${amount}`);
	}, [location]);

	const handleProductChange = useCallback(async ({currentTarget}) => {
		const locationId = state?.location?.id;

		const { value } = currentTarget;
		const product = products.find(v => v.id === Number(value));

		if (product) {
			const result = await post(`/${locationId}/product/${product.id}`);

			if (result) {
				setProduct(product);

				setState(prev => {
					const state = {...prev, standard: result};
					console.log(`${state.location.name}: set product standard to ${product.name}`)
					console.log(result);
					return state;
				});

				setProductLock(true);
			}
		} else {
			console.log(`Selected product ${value} not found in`, products);
		}
	}, [products, state]);

	const handleFilterChange = useFilterChangeCallback(shift, shifts, (i) => setShift(i));

	const { indicators } = state;
	const { availability, cycleTime, oee, performance, quality, status, totalPieces } = indicators || {};
	const { interval } = lineProcess;
	let dateRangeInterval = [];
	if (interval && interval.range) {
		dateRangeInterval = interval.range.map(s => new Date(s)) as TimeRange;
	}
	let [start, end] = ['' , '']
	if (dateRangeInterval.length > 1)
		[start, end] = dateRangeInterval;

	return <Fragment>
		<div className="filter-container">
			<TimeFilter
				date={date}
				timeFilter={timeFilter}
				shiftFilter={String(Number(shift?.id) || 0)}
				shiftFilters={shifts}
				onChange={handleFilterChange}
			/>
		</div>

		<div className="asd">
			<h3>
				{state.location?.name} · <span>Producing:</span>
			</h3>

			<Select
				outlined
				className="dense"
				disabled={!!state.standard && productLock}
				value={String(product?.id) || ""}
				placeholder="Product"
				onChange={handleProductChange}
			>
				{products.map(({id, name}: Product) => <option key={id} value={id}>{name}</option>)}
			</Select>

			{
				(realtime && timeFilter === "Shift" && state.standard && productLock) &&
				<IconButton
					className="date-button"
					icon="lock"
					title="Unlock"
					onClick={() => setProductLock(false)}
				/>
			}
			
			<h3>
				{realtime && timeFilter === "Shift"  && <span>· realtime </span>}
			</h3>
			&nbsp; 
			&nbsp; 
			<h3>
				{<span> ·  Range:  {moment(start).format("MM/DD/YYYY hh:mm A")}  -  {moment(end).format("MM/DD/YYYY hh:mm A")}</span> }
			</h3>
		</div>

		<IndicatorCard
			subtitle="Availability"
			title={`${availability || 0}%`}
			background="#3c8dbc"
			icon="clock-outline"
		/>

		<IndicatorCard
			subtitle="Performance"
			title={`${performance || 0}%`}
			background="#3c8dbc"
			icon="calculator"
		/>

		<IndicatorCard
			subtitle="Quality"
			title={`${quality || 0}%`}
			background="#3c8dbc"
			icon="calculator"
			onClickOptions={realtime && timeFilter === "Shift" && {
				content: "Add scrap info",
				handle: () => setOpen(true)
			}}
		/>

		<IndicatorCard
			subtitle={<StandardSubtitle text="OEE" stdValue="90%"/>}
			title={`${oee || 0}%`}
			background={calcColor(oee, 90)}
			icon="calculator"
		/>

		<IndicatorCard
			subtitle={
				state.standard ?
					<StandardSubtitle
						text="Pipe Count"
						stdValue={
							state.standard ?
								`${state.standard.value} ${state.standard.unit}` :
								"-"
						}
					/> :
					"Pipe Count"
			}
			title={totalPieces ? totalPieces : "-"}
			background={calcColor(totalPieces || 0, 1)}
			icon="calculator"
		/>

		<IndicatorCard
			subtitle={
				state.standard ?
					<StandardSubtitle
						text="Cycle Time"
						stdValue={state.standard ? (1 / state.standard.value * 3600).toFixed(2) + " sec" : "-"}
					/> :
					"Cycle Time"
			}
			title={cycleTime ? `${(Number(cycleTime) / 1000).toFixed(1)} sec` : "-"}
			background={calcColor(cycleTime, 1 / state.standard?.value * 3600)}
			icon="clock-outline"
		/>

		{
			realtime &&
			<IndicatorCard
				subtitle="Line Status"
				title={status || "No conn."}
				background={calcColor(status, "Running")}
				icon="chart-bar"
			/>
		}

		<ScrapDialog
			open={open}
			product={product}
			renderToPortal
			onClose={({detail: {action}}) => {
				console.log(action);
				setOpen(false);
			}}
			onAddClick={handleAddClick}
			// onClosed={evt => console.log(evt.detail.action)}
		/>
	</Fragment>;
}

interface ScrapDialogProps extends DialogProps {
	product?: Product;
	amount?: number;
	onAddClick: (value: number) => void;
}

const ScrapDialog: FC<ScrapDialogProps> = ({product, amount, onAddClick, ...props}) => {
	const [scrap, setScrap] = useState<number>(0);

	useEffect(() => setScrap(amount || 0), [amount]);

	const handleAddClick = useCallback(() => {
		setScrap(prev => {
			onAddClick(prev);
			return prev;
		})
	}, []);

	return <>
		<Dialog {...props}>

			{(product && <>
				<DialogTitle>
					Add scrap of {product.name}
				</DialogTitle>

				<DialogContent>
					<TextField
						placeholder="Amount of pieces"
						type="number"
						value={scrap}
						onChange={({ currentTarget: { value } }) => {
							const amount = Number(value || 0);
							setScrap(amount >= 0 ? amount : 0);
						}}
					/>
				</DialogContent>
			</>) || <>
				<DialogTitle>
					Action required first
				</DialogTitle>

				<DialogContent>
					You should first set which product is being produced in this line.
				</DialogContent>
			</>}

			<DialogActions>
				<DialogButton action="close">
					Cancel
				</DialogButton>

				{
					product &&
					<DialogButton action="add" onClick={handleAddClick}>
						Add
					</DialogButton>
				}
			</DialogActions>
		</Dialog>
	</>;
}