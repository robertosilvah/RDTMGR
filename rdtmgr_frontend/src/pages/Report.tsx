import React, { FC, useContext, useEffect, useState } from "react";

import moment from "moment";

import { ParentSize } from "@visx/responsive";

import { Card } from "rmwc";

import { Gantt } from "../components/reports/Gantt";
import { IndicatorsCard } from "../components/reports/Indicators";
import { FullscreenSpinner } from "../components/Spinners";
import TimeFilter, { TimeInterval } from "../components/filters/TimeFilter";
import { getShiftsStart, useFilterChangeCallback } from "../CustomHooks";
import { FilterContext, FilterProvider, LineProcessContext, LineProcessesContext, LineProcessProvider, TimeRange } from "../Context";

import "./Report.scss";

export const Report = () => {
	const lineProcesses = useContext(LineProcessesContext);

	return (
		<div className="report-content">
			<FullscreenSpinner/>

			{
				lineProcesses ?
					Array.from(lineProcesses.values()).map(process => (
						<FilterProvider key={process.location.name}>
							<LineProcessProvider lineProcess={process}>
								<LineProcessComponent />
							</LineProcessProvider>
						</FilterProvider>
					)) : "Loading..."
			}
		</div>
	);
}

const shouldShowGantt = (timeFilter) => {
	switch (timeFilter) {
		case "Week":
		case "Month":
			return false;
		default:
			return true;
	}
}

const LineProcessComponent: FC = () => {
	const { date, dateRange, timeFilter, setFilters } = useContext(FilterContext);

	const lineProcess = useContext(LineProcessContext);

	const [showGantt, setShowGantt] = useState(shouldShowGantt(timeFilter));
	const [shift, setShift] = useState<TimeInterval>();

	const { shifts, realtime } = lineProcess;
	// const [realtime, setRealtime] = useState(true);

	// updates the state and the date range if it changed
	useEffect(() => {
		if (lineProcess && realtime && timeFilter === "Shift") {
			if (shift?.id !== lineProcess.interval?.id) {
				const { interval } = lineProcess;

				if (interval) {
					console.log(`${lineProcess.location?.name}: updating shift to`, interval);
					const dateRange = interval.range.map(s => new Date(s)) as TimeRange;

					setFilters({ dateRange, shiftFilter: interval.id, date: getShiftsStart(shifts) });
					setShift(interval);
				}
			}
		}
	}, [lineProcess, realtime, shifts]);

	// toggle gantt chart
	useEffect(() => {
		setShowGantt(shouldShowGantt(timeFilter));
	}, [timeFilter]);

	const handleFilterChange = useFilterChangeCallback(shift, shifts, (i) => setShift(i));

	const { lineStops, delayTimeStr, delayRate, oee, status, totalPieces, gantt } = lineProcess?.indicators || {};
	const shiftPosition = String(Number(shift?.id) || 0);
	const { interval } = lineProcess;
	let dateRangeInterval = [];
	if (interval && interval.range) {
		dateRangeInterval = interval.range.map(s => new Date(s)) as TimeRange;
	}
	let [start, end] = ['' , '']
	if (dateRangeInterval.length > 1)
		[start, end] = dateRangeInterval;

	return <>
		<div className="filter-container">
			<TimeFilter
				date={date}
				timeFilter={timeFilter}
				shiftFilter={shiftPosition}
				shiftFilters={shifts}
				onChange={handleFilterChange}
			/>
		</div>

		<IndicatorsCard
			line={<>
				{lineProcess.location?.name} {realtime && timeFilter === "Shift" && <span style={{ opacity: .5 }}>· realtime </span>}&nbsp; {<span style={{ opacity: .5 }}> ·  Range:  {moment(start).format("MM/DD/YYYY hh:mm A")}  -  {moment(end).format("MM/DD/YYYY hh:mm A")}</span> }
			</>}
			entry={totalPieces}
			stops={lineStops}
			delayTime={delayTimeStr}
			delayRate={delayRate}
			efficiency={oee}
			lineStatus={status}
			showLineStatus={realtime && timeFilter === "Shift" }
		/>

		{
			showGantt &&
			<Card>
				<ParentSize style={{ padding: "16px 16px 8px" }}>
					{({ width }) => <Gantt
						width={width}
						data={gantt}
						dateRange={dateRange}
						format={timeFilter !== "Shift" ? "%H" : "%H:%M"}
					/>}
				</ParentSize>
			</Card>
		}
	</>;
}

export default Report;
