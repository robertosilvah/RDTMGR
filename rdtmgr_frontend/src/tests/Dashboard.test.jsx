import '@testing-library/dom';

import React from 'react';
import { Dashboard } from '../pages/Dashboard';
import { FilterContext } from '../Context';
import { AppContext } from '../Context';
import { Shallow } from './setupTest'

describe('Test in <DashBoard />', () => {
	test ('No 1', () => {
		const context = {
			api: {
				main: process.env.REACT_APP_API,
				webSocket: process.env.REACT_APP_WS
			}
		};

		const filter = {
			date: new Date(),
			dateRange: [new Date(), new Date()+1],
			timeFilter: 'Shift',
			shiftFilter: 'Third',
			weekFilter: null,
			search: undefined,
			changed: false,
			shiftFilters: null,
			setFilters: null
		};

		const wrapper = Shallow(
			<AppContext.Provider value = {context}>
				<FilterContext.Provider value= {filter}>
					<Dashboard />
				</FilterContext.Provider>
			</AppContext.Provider>
		);

		expect(wrapper).toMatchSnapshot();
	})
})