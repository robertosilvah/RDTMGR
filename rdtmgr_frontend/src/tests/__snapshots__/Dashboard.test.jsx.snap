// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <DashBoard /> No 1 1`] = `
<ContextProvider
  value={
    Object {
      "changed": false,
      "date": 2021-02-19T19:01:07.439Z,
      "dateRange": Array [
        2021-02-19T19:01:07.439Z,
        "Fri Feb 19 2021 16:01:07 GMT-0300 (Argentina Standard Time)1",
      ],
      "search": undefined,
      "setFilters": null,
      "shiftFilter": "Third",
      "shiftFilters": null,
      "timeFilter": "Shift",
      "weekFilter": null,
    }
  }
>
  <Dashboard />
</ContextProvider>
`;
